bentoml/frameworks/paddle.py

```diff
@@ -0,0 +1,103 @@
+import os
+
+from bentoml.exceptions import MissingDependencyException
+from bentoml.service.artifacts import BentoServiceArtifact
+from bentoml.service.env import BentoServiceEnv
+import tempfile
+
+
+try:
+    import paddle
+    import paddle.inference as paddle_infer
+except ImportError:
+    paddle = None
+
+
+class PaddleModelArtifact(BentoServiceArtifact):
+    """Abstraction for saving/loading PaddlePaddle models
+
+    Args:
+        name (string): name of the artifact
+
+    Raises:
+        MissingDependencyException: paddle package is required for PaddleModelArtifact
+
+    Example usage:
+
+    >>> import pandas as pd
+    >>>
+    >>> from bentoml import env, artifacts, api, BentoService
+    >>> from bentoml.adapters import DataframeInput
+    >>> from bentoml.frameworks.paddle import PaddleModelArtifact
+    >>>
+    >>> @env(infer_pip_packages=True)
+    >>> @artifacts([PaddleModelArtifact('model')])
+    >>> class PaddleService(BentoService):
+    >>>    @api(input=DataframeInput(), batch=True)
+    >>>    def predict(self, df: pd.DataFrame):
+    >>>        input_data = df.to_numpy().astype(np.float32)
+    >>>        predictor = self.artifacts.model
+    >>>
+    >>>        input_names = predictor.get_input_names()
+    >>>        input_handle = predictor.get_input_handle(input_names[0])
+    >>>        input_handle.reshape(input_data.shape)
+    >>>        input_handle.copy_from_cpu(input_data)
+    >>>
+    >>>        predictor.run()
+    >>>
+    >>>        output_names = predictor.get_output_names()
+    >>>        output_handle = predictor.get_output_handle(output_names[0])
+    >>>        output_data = output_handle.copy_to_cpu()
+    >>>        return output_data
+    >>>
+    >>> service = PaddleService()
+    >>>
+    >>> service.pack('model', model_to_save)
+    """
+
+    def __init__(self, name: str):
+        super(PaddleModelArtifact, self).__init__(name)
+        self._model = None
+        self._predictor = None
+        self._model_path = None
+
+        if paddle is None:
+            raise MissingDependencyException(
+                "paddlepaddle package is required to use PaddleModelArtifact"
+            )
+
+    def pack(self, model):  # pylint:disable=arguments-differ
+        self._model = model
+
+        if self._model_path is None:
+            self._model_path = tempfile.TemporaryDirectory().name
+            self._save(self._model_path)
+
+        config = paddle_infer.Config(
+            self._model_path + ".pdmodel", self._model_path + ".pdiparams"
+        )
+        config.enable_memory_optim()
+        predictor = paddle_infer.create_predictor(config)
+        self._predictor = predictor

```

Comment:
Maybe we can move these code to  the `get` function? That way we don't necessarily need to create `predictor` if the users don't run predictions before save.

See `OnnxModelArtifact` for similar pattern

Reply:
Correct me if I'm misunderstanding but I thought an artifact needs to be packed with a model before it can be used for inference? Is there a case where an artifact runs predictions without packing a model?

Reply:
I see what you're saying. If I'm not mistaken, `get` is called everytime the user run a prediction correct? So if the user runs multiple predictions, the `predictor` would be unnecessarily initialized multiple times.

Reply:
I can add a check if self._predictor already exists maybe to prevent the issue mentioned?